src/beamlime/constructors/inspectors.py

```diff
@@ -52,9 +66,9 @@ class ProductSpec:
     Specification of a product (returned value) of a provider.
     """
 
-    def __init__(self, product_type: Union[ProductType, ProductSpec]) -> None:
-        self.product_type: ProductType
-        self.returned_type: ProductType
+    def __init__(self, product_type: Union[Type[Product], ProductSpec]) -> None:
+        self.product_type: Type[Product]
+        self.returned_type: Type[Product]
```

Comment:
I wanted to keep them bound in the `ProviderGroup.__getitem__`.
And then whenever I annotate `product_type` with `type[any]`,
static type checker complains that it's not `type[Product]`...
And since the `returned_type` is the type of what is actually returned from a callable,
it is should always be a super class of `product_type` so I think this annotation makes sense...?
```
class NewInt(int): ...
product = NewInt(1)
```
With the case above,
```
isinstance(product, int)
isinstance(product, NewInt)
```

`returned_type` will be `int` and `product_type` will be `NewInt`,
except that it'll be created by `NewType` instead of inheriting `int`...?

And if it's not `NewType`, they will be exact same thing.